pkg/controllers/leaderworkerset_controller.go

```diff
@@ -537,51 +543,101 @@ func (r *LeaderWorkerSetReconciler) iterateReplicas(ctx context.Context, lws *le
-		if !(podTemplateHash == revisionKey && podutils.PodRunningAndReady(sortedPods[index])) {
-			return false
-		}
+		leaderUpdated := revisionutils.GetRevisionKey(&sortedPods[idx]) == revisionKey
+		leaderReady := podutils.PodRunningAndReady(sortedPods[idx])
 
 		if noWorkerSts {
-			return true
+			states[idx] = replicaState{
+				ready:   leaderReady,
+				updated: leaderUpdated,
+			}
+			continue
+		}
+
+		stsExists := (int(idx) < len(sortedSts) && nominatedName == sortedSts[idx].Name)
+		if !stsExists {
+			continue
 		}
 
-		stsTemplateHash := revisionutils.GetRevisionKey(&sortedSts[index])
-		return stsTemplateHash == revisionKey && statefulsetutils.StatefulsetReady(sortedSts[index])
+		workersUpdated := revisionutils.GetRevisionKey(&sortedSts[idx]) == revisionKey
+		workersReady := statefulsetutils.StatefulsetReady(sortedSts[idx])
+
+		states[idx] = replicaState{
+			ready:   leaderReady && workersReady,
+			updated: leaderUpdated && workersUpdated,
+		}
 	}
 
-	var skip bool
-	var continuousReadyReplicas, lwsUnreadyReplicas int32
+	return states, nil
+}
 
-	for index := stsReplicas - 1; index >= 0; index-- {
-		replicaReady := processReplica(index)
-		skip = skip || !replicaReady
+func calculateLWSUnreadyReplicas(states []replicaState, lwsReplicas int32) int32 {
+	var unreadyCount int32
+	for idx := int32(0); idx < min(int32(len(states)), lwsReplicas); idx++ {
+		if !states[idx].ready || !states[idx].updated {
+			unreadyCount++
+		}
+	}
+	return unreadyCount
+}
 
-		if replicaReady && !skip {
-			continuousReadyReplicas++
+func calculateContinuousReadyReplicas(states []replicaState) int32 {
+	// Count ready replicas at tail (from last index down)
+	var continuousReadyCount int32
+	for idx := len(states) - 1; idx >= 0; idx-- {
+		if !states[idx].ready || !states[idx].updated {
+			break
 		}
-		if !replicaReady && index < *lws.Spec.Replicas {
-			lwsUnreadyReplicas++
+		continuousReadyCount++
+	}
+	return continuousReadyCount
+}
+
+// Checks state of not updated replicas and adjusts the rolling update parittion to prevent
+// availability drops below "maxUnavailable".
+func partitionAccountingForUnavailable(states []replicaState, desiredPartition int32) int32 {
+	// Desired partition already accounts for readiness of replicas above partition. Code below
+	// only needs to consider replicas below desired partition.
+	var unavailableBelowPartition int32
+	for idx := 0; idx < int(desiredPartition); idx++ {
+		if !states[idx].ready {
+			unavailableBelowPartition++
+		}
+	}
+
+	partitionAccountingForUnavailable := desiredPartition + unavailableBelowPartition
+
+	// Reduce the partitionAccountingForUnavailable as long as replicas are unavailable. This is safe to update these
+	// replicas as it does not impact general availability.
+	// Never reduce below desiredPartition.
+	for idx := partitionAccountingForUnavailable; idx >= desiredPartition; idx-- {
+		if idx < int32(len(states)) && states[idx].ready {
```

Comment:
I think we're playing a trick here, the first ready partition is actually not the accurate one, which may occur like the partition is getting bigger and bigger, but as mentioned, it won't impact the final result.

Another approach is comparing the current partition with the original one, but we need to pay attention to partition=0.